vllm/model_executor/models/ultravox.py

```diff
@@ -140,78 +92,99 @@ def input_mapper_for_ultravox(ctx: InputContext, data: object):
-            "multi_modal_placeholders"]:
-        # The inputs already have placeholders.
-        return inputs
+    def _apply_hf_processor(
+        self,
+        prompt: str,
+        mm_data: MultiModalDataDict,
+        mm_processor_kwargs: Mapping[str, object],
+    ) -> BatchFeature:
+        audio_data = mm_data["audio"]
+        if not isinstance(audio_data, list):
+            audio_data = [audio_data]
+
+        # Ultravox processor doesn't support multiple inputs,
+        # therefore we need to input text and audio one by one
+        tokenizer = self._get_tokenizer()
+        hf_inputs = defaultdict(list)
+        for audio, sr in audio_data:
+            data = self._resample_audio(audio, sr)
+            processed_inputs = super()._apply_hf_processor(
+                prompt, data, mm_processor_kwargs)
+            prompt = tokenizer.decode(processed_inputs["input_ids"][0],
+                                      skip_special_tokens=False)
+            hf_inputs["audio_features"].append(
+                processed_inputs["audio_values"].squeeze(0))
+        hf_inputs["input_ids"] = processed_inputs["input_ids"]
+        return hf_inputs
+
+    def _get_hf_processor(
+        self,
+        **mm_processor_kwargs: Mapping[str, object],
+    ) -> ProcessorMixin:
+        # Ultravox processor use eot_token_id as the audio placeholder token,
+        # we replace it with <|reserved_special_token_0|> for convenience.
+        hf_processor = super()._get_hf_processor(**mm_processor_kwargs)
+        hf_processor.audio_token_replacement = _AUDIO_PLACEHOLDER_STR
+        return hf_processor
+
+    def _get_processor_data(
+        self,
+        mm_data: MultiModalDataDict,
+    ) -> Tuple[Dict[str, Any], Dict[str, Any]]:
+        # Ultravox uses "audio" instead of "audios" as calling keyword
+        processor_data, passthrough_data = super()._get_processor_data(mm_data)
+        if "audios" in processor_data:
+            processor_data["audio"] = processor_data.pop("audios")
+        return processor_data, passthrough_data
+
+    def _get_prompt_replacements(
+        self,
+        mm_items: MultiModalDataItems,
+        hf_inputs: BatchFeature,
+        mm_processor_kwargs: Mapping[str, object],
+    ) -> list[PromptReplacement]:
+        hf_processor = self._get_hf_processor()
+        stack_factor = hf_processor.stack_factor
+        encoder_ds_factor = hf_processor.encoder_ds_factor
+
+        def get_replacement_ultravox(item_idx: int):
+            audio_data, sr = mm_items.audio[item_idx]
+            audio_data = self._resample_audio(audio_data, sr)["audio"]
+            audio_len = audio_data.shape[0]
+            nb_encoder_frames = int(round(audio_len / encoder_ds_factor +
+                                          1e-4))
+            audio_token_len = int(np.ceil(nb_encoder_frames / stack_factor))
+            max_audio_token_len = get_ultravox_max_audio_tokens(self.ctx)
```

Comment:
```suggestion
            max_audio_token_len = get_ultravox_max_audio_tokens(self.ctx)
```

We can move this line outside of the local function.